server/catgenome/src/main/java/com/epam/catgenome/util/feature/reader/TribbleIndexedFeatureReader.java

```diff
@@ -0,0 +1,629 @@
+
+                        currentRecord = f;     // Success
+                        return;
+
+                    } catch (TribbleException e) {
+                        e.setSource(path);
+                        throw e;
+                    } catch (NumberFormatException e) {
+                        final String error = "Error parsing line: " + source;
+                        throw new TribbleException.MalformedFeatureFile(error, path, e);
+                    }
+                }
+                if (blockIterator != null && blockIterator.hasNext()) {
+                    advanceBlock();   // Advance to next block
+                } else {
+                    return;   // No blocks left, we're done.
+                }
+            }
+        }
+
+
+        public void remove() {
+            throw new UnsupportedOperationException("Remove is not supported.");
+        }
+
+
+        public void close() {
+            // Note that this depends on BlockStreamWrapper not actually closing the underlying stream
+            codec.close(source);
+            if (!reuseStreamInQuery()) {
+                // if we are going to reuse the underlying stream we don't close the underlying stream.
+                try {
+                    mySeekableStream.close();
+                } catch (IOException e) {
+                    throw new TribbleException("Couldn't close seekable stream", e);
+                }
+            }
+        }
+
+        public Iterator<T> iterator() {
+            return this;
+        }
+    }
+
+
+    /**
+     * Wrapper around a SeekableStream that limits reading to the specified "block" of bytes.  Attempts to
+     * read beyond the end of the block should return -1  (EOF).
+     */
+    static class BlockStreamWrapper extends InputStream {
+
+        SeekableStream seekableStream;
+        long maxPosition;
+
+        BlockStreamWrapper(final SeekableStream seekableStream, final Block block) throws IOException {
+            this.seekableStream = seekableStream;
+            seekableStream.seek(block.getStartPosition());
+            maxPosition = block.getEndPosition();
+        }
+
+        @Override
+        public int read() throws IOException {
+            return (seekableStream.position() > maxPosition) ? -1 : seekableStream.read();
+        }
+
+        @Override
+        public int read(final byte[] bytes, final int off, final int len) throws IOException {
+            // note the careful treatment here to ensure we can continue to
+            // read very long > Integer sized blocks
+            final long maxBytes = maxPosition - seekableStream.position();
+            if (maxBytes <= 0) {
+                return -1;
+            }
+
+            final int bytesToRead = (int) Math.min(len, Math.min(maxBytes, Integer.MAX_VALUE));
+            return seekableStream.read(bytes, off, bytesToRead);
+
+        }
+    }
+
+    protected static class IndexCache <T extends Feature, S> implements CacheIndex {
```

Comment:
Please rename it to reflect it's purpose: agreed to rename this to TribbleIndexCache